kmod/src/omap.c

```diff
@@ -0,0 +1,977 @@
+			call_rcu(&req->rcu, free_req_rcu);
+		}
+	}
+	rcu_read_unlock();
+
+	if (send_response)
+		ret = scoutfs_server_send_omap_response(sb, resp_rid, resp_id, map, 0);
+	else
+		ret = 0;
+	kfree(map);
+out:
+	return ret;
+}
+
+/*
+ * The server is shutting down.  Free all the server state associated
+ * with ongoing request processing.  Clients who still have requests
+ * pending will resend them to the next server.
+ */
+void scoutfs_omap_server_shutdown(struct super_block *sb)
+{
+	DECLARE_OMAP_INFO(sb, ominf);
+	struct rhashtable_iter iter;
+	struct llist_node *requests;
+	struct omap_request *req;
+	struct omap_request *tmp;
+
+	rhashtable_walk_enter(&ominf->req_ht, &iter);
+	rhashtable_walk_start(&iter);
+
+	for (;;) {
+		req = rhashtable_walk_next(&iter);
+		if (req == NULL)
+			break;
+		if (req == ERR_PTR(-EAGAIN))
+			continue;
+
+		if (req->rids.nr_rids != 0) {
+			free_rids(&req->rids);
+			rhashtable_remove_fast(&ominf->req_ht, &req->ht_head, req_ht_params);
+			call_rcu(&req->rcu, free_req_rcu);
+		}
+	}
+
+	rhashtable_walk_stop(&iter);
+	rhashtable_walk_exit(&iter);
+
+	requests = llist_del_all(&ominf->requests);
+	llist_for_each_entry_safe(req, tmp, requests, llnode)
+		kfree(req);
+
+	synchronize_rcu();
+}
+
+/*
+ * Return 1 and give the caller a write inode lock if it is safe to be
+ * deleted.  It's safe to be deleted when it is no longer reachable and
+ * nothing is referencing it.
+ *
+ * The inode is unreachable when nlink hits zero.  Cluster locks protect
+ * modification and testing of nlink.  We use the ino_lock_cov covrage
+ * to short circuit the common case of having a locked inode that hasn't
+ * been deleted.  If it isn't locked, we have to acquire the lock to
+ * refresh the inode to see its current nlink. 
+ *
+ * Then we use an open inode bitmap that covers all the inodes in the
+ * lock group to determine if the inode is present in any other mount's
+ * caches.  We refresh it by asking the server for all clients' maps and
+ * then store it in the lock.  As long as we hold the lock nothing can
+ * increase nlink from zero and let people get a reference to the inode.
+ */
+int scoutfs_omap_should_delete(struct super_block *sb, struct inode *inode,
+			       struct scoutfs_lock **lock_ret)
+{
+	struct scoutfs_inode_info *si = SCOUTFS_I(inode);
+	struct scoutfs_lock *lock = NULL;
+	const u64 ino = scoutfs_ino(inode);
+	struct omap_lock *olk;
+	u64 group_nr;
+	int bit_nr;
+	u64 vers;
+	int ret;
+
+	/* lock group and omap constants are defined independently */
+	BUILD_BUG_ON(SCOUTFS_OPEN_INO_MAP_BITS != SCOUTFS_LOCK_INODE_GROUP_NR);
+
+	if (scoutfs_lock_is_covered(sb, &si->ino_lock_cov) && inode->i_nlink > 0) {
+		ret = 0;
+		goto out;
+	}
+
+	ret = scoutfs_lock_inode(sb, SCOUTFS_LOCK_WRITE, SCOUTFS_LKF_REFRESH_INODE, inode, &lock);
+	if (ret < 0)
+		goto out;
+
+	if (inode->i_nlink > 0) {
+		ret = 0;
+		goto out;
+	}
+
+	/* multiple callers can be all trying to allocate an omap_lock for the lock */
```

Comment:
Yeah, I'm glad you called this out.  It started being derived from a similar pattern over in dentry handling.  That is in a very hot path and only has the dentry's d_private long to serialize with so it gets all bonkers.  Having started with that aggressive pattern I then kept running with it long after it became subtle and dense.   As you point out, it can be a lot simpler.  We can add a real lock to our lock struct for omap to work with and this isn't a really hot path so the dumb reads-are-stores of spinlocks is meh, fine.  I reworked the two little operations into much clearer functions.  Hopefully that's better.